### ISSUE TITLE
Problem in resetting choco solver/model

### ISSUE THREAD

**Issue by ritika2734** [chocoteam/choco-solver#549]
Hi,

I am using choco solver 4. I am constructing a tree with choco model at each node. For some purpose I need to make multiple copies of the  same tree having choco model. within nodes  When I solve the first tree, by providing some constraints it gives me correct solution. However, for next tree,  when I change the constraints, the solver does not provide the correct solution. How can I reset the solver as well as the complete model so that it can execute from the beginning?

I used solver.reset() but it does not work either.

**Comment by cprudhom** [comment]
I think the problem is fixed in the last [snapshot release](https://oss.sonatype.org/content/repositories/snapshots/org/choco-solver/choco-solver/4.0.5-SNAPSHOT/).
If that's not the case, please provide us a minimal working example we can execute.

Best regards,

**Comment by ritika2734** [comment]
Okay. Thank you for the reply . I will test it and let you know, if I face any problem.

**Comment by ritika2734** [comment]
Hi, 
I tried to reset the solver using latest release with "solver.reset()" command but it does not work.
Also, I need to reset the variables value.Example:
       // 1. Create a Model
        Model model = new Model("my first model");

        // 2. Create variables
        BoolVar a = model.boolVar("a");
        BoolVar b = model.boolVar("b");

        // 3. Post constraints
        model.arithm(a,"=",1).post();
        model.arithm(b,"=",1).post();
        model.addClauses(and(a, nor(b)));

        // 4. Solve the problem
        model.getSolver().solve(); 
        if(!solver.findAllSolutions().isEmpty()){
					setBooleanValue(true);
				}else{
					setBooleanValue(false);
				}
solver.reset();

Again, I want to solve the model with different constraints like:

 // 3. Post constraints
        model.arithm(a,"=",0).post();
        model.arithm(b,"=",0).post();
        model.addClauses(or(a, nor(b)));
        model.getSolver().solve();

**Comment by cprudhom** [comment]
Could you clarify what you're excepting the `reset()` method to do ?

Resetting a model does not erase the constraints and in your case, you are posting contradictory constraints (a = 1 then a = 0).

Finally, clauses are managed by a third- party library and there is no way to remove a clause.

**Comment by ritika2734** [comment]
Hi, 

I want to reset the constraints value as I want to use the same choco model with same clause to give me different results when I use different constraints. Please find attached the example. Hope it helps
[Choco model for different constraints.pdf](https://github.com/chocoteam/choco-solver/files/1324103/Choco.model.for.different.constraints.pdf)
.

**Comment by cprudhom** [comment]
> I want to use the same choco model with same clause
But you posted two different clauses : 
`model.addClauses(and(a, nor(b)))` and `model.addClauses(or(a, nor(b)))`
(Btw, is it a `nor` or a `not`?)

> I want to use the same choco model with same clause to give me different results when I use different constraints
If there are different constraints from resolution to the other then models are not the same.
Maybe you have a 'core' model with variables and hard constraints (for example, that two features are exclusive).
Based on this core model, you add additional constraints to "try" configurations.
If those additional constraints are unary constraints, you should consider using a search strategy and find all solutions that match the core model.

Finally, if models are really different, you should consider creating a new one each time.

**Comment by ritika2734** [comment]
Hi,

I have attached one pdf in the above comment for your reference. The clause remain same in every node only the value of the variables needs to be reset.

**Comment by ritika2734** [comment]
Also, there is no "not" operator so I have used "nor"  instead.

**Comment by cprudhom** [comment]
> The clause remain same in every node

So why posting an `and` and then a `or` ?

There is a `not` operator (it is names 'negate' which is not very common, I agree) but you can also call `b.not()`

If only the values change, you should use a pre-defined (or your own) search strategy: that's the aim of a solver to find assignments that satisfy the constraints, so you don't have to do it manually by adding/removing constraint.

**Comment by ritika2734** [comment]
Hello,

Thank you for your comment. I have attached a pdf with my previous comment .I think you have not seen the pdf that I have attached. I am attaching the pdf again for your reference.
[Choco.model.for.different.constraints.pdf](https://github.com/chocoteam/choco-solver/files/1324649/Choco.model.for.different.constraints.pdf) . Also, I will try to find the solution with the search strategy and let you know.

**Comment by ritika2734** [comment]
Hello​ Team,

I tried to find and implement the search stratergy but it does not work. I will try to rephrase my question hope that this time I am able to explain it properly.
I have a choco model and a constraint(LogOp) attached to it. 
I want to apply the constraint from the user which is present in the "productFeatureSet" to see whether the constraint(LogOp) satisfies it.
The user can give different values in the "productFeatureSet" and I need to apply these values again and again to same model and find solution.

Code:
Case 1
LogOp tree= LogOp.and(Feature1,Feature2); //Constraint
org.chocosolver.solver.Model chocoModel=this.getChocoModelField(); // Choco Model
productFeatureSet=[Feature1]	//Provided by user, need to set its value to 1
if(tree!=null){
    for(BoolVarImpl var: this.getChocoVariablesList()){
        if(productFeatureSet.contains(var.getName())){
            chocoModel.arithm(var,"=",1).post(); 
        }else{
            chocoModel.arithm(var,"=",0).post();
        }
    }
    chocoModel.addClauses(tree);

}else{
    chocoModel.trueConstraint();	
}


Solver solver= chocoModel.getSolver();
Solution s= solver.findSolution();
if(s!=null){
    System.out.println("Solution exists");
}else{
    System.out.println("No Solution ");
}

Here, in case1 I want Feature1=1 and Featur2=0
​
Case 2:
Again, user provide new productFeatureSet=[Feature2] for same model and I need to test the condition Feature2=1 and Feature1=0 on same LogOp constraint. However, the model is not able to show the correct result this time.

Please let me know if you need more information.

**Comment by cprudhom** [comment]
Ok, let me rephrase to be sure I understand well.

1. You have a model *M1* composed of variables and constraints, which represents a problem.
The model itself is feasible, it has at least one solution.
2. A user makes some choices over variables of *M1* which is transformed into a constraint *Cu*.
3. What you want to check is *M1+Cu* is still feasible and if true, exhibit a solution.
4. In case *M1+Cu* has no solution, *Cu* is removed, and go to 2.

ok ?

**Comment by ritika2734** [comment]
Hi,

Yes, Model M1 has a LogOp constraint associated to it and it is fixed. The user makes the variable choices by selecting any variables and initialising the selected values to 1 and then evaluating it against that LopOp constraint. 
 The user can again pick the new values and test it against the fixed constraint.  This process can occur "n" number of times.

**Comment by cprudhom** [comment]
Here is a proposal, I hope it is commented enough to be self-content.
If not, tell me.


```java
// --- core model                                                                    
Model model = new Model();                                                           
BoolVar[] features = model.boolVarArray("F", 2);                                     
LogOp tree= LogOp.and(features[0],features[1]); //Constraint                         
model.addClauses(tree);                                                              
int[][] requests = new int[][]{                                                      
        {1,0}, // F1 but not F2 -> no solution                                       
        {-1,1}, // nothing on F1 and F2-> a solution                                 
        {1,1}, // F1 + F2 -> a solution                                              
        // ...                                                                       
};                                                                                   
                                                                                     
List<Constraint> temp = new ArrayList<>(); // to be able to remove user's request    
for(int i = 0 ; i< requests.length; i++){                                            
    model.getEnvironment().worldPush(); // to be able to restore initial state       
    // transform user's request into constraints                                     
    for(int j = 0; j < requests[i].length; j++){                                     
        switch (requests[i][j]){                                                     
            case -1:                                                                 
                break;                                                               
            case 0:                                                                  
                Constraint c0 = model.arithm(features[j], "=", 0);                   
                temp.add(c0);                                                        
                c0.post();                                                           
                break;                                                               
            case 1:                                                                  
                Constraint c1 = model.arithm(features[j], "=", 1);                   
                temp.add(c1);                                                        
                c1.post();                                                           
                break;                                                               
        }                                                                            
    }                                                                                
    // now try to find a solution                                                    
    Solver solver = model.getSolver();                                               
    solver.hardReset();                                                              
    if(solver.solve()){                                                              
        System.out.println("Solution exists");                                       
        System.out.printf("%d %d\n", features[0].getValue(),features[1].getValue()); 
    } else {                                                                         
        System.out.println("No Solution ");                                          
    }                                                                                
    // restore initial state                                                         
    model.getEnvironment().worldPopUntil(0);                                         
    // remove all constraints relative to user's request                             
    temp.forEach(model::unpost);                                                     
    temp.clear();                                                                    
}                                                                                    
```

**Comment by ritika2734** [comment]
Hello ,

I can have multiple "n" choco variables associated with the model. For the example purpose, I just provided 2. Is there a way in choco to create these requests array or I have to do it by code? 

Also, Thanks for such a quick response.

**Comment by ritika2734** [comment]
`chocoModel.addClauses(tree);
List<Constraint> temp = new ArrayList<>(); 
 for(BoolVarImpl var: this.getChocoVariablesList()){
			chocoModel.getEnvironment().worldPush(); // to be able to restore initial state 
						if(productFeatureSet.contains(var.getName())){
							//chocoModel.arithm(var,"=",1).post();
					Constraint c1 = chocoModel.arithm(var, "=", 1);                   
			                temp.add(c1);                                                        
			                c1.post(); 
						}else{
					//chocoModel.arithm(var,"=",0).post();
					Constraint c0 = chocoModel.arithm(var, "=", 0);                   
			                temp.add(c0);                                                        
			                c0.post(); 
						}
					}
					Solver solver = chocoModel.getSolver();                                               
				    solver.hardReset();                                                              
				    if(solver.solve()){                                                              
				        System.out.println("Solution exists");  
				       
				    } else {                                                                         
				        System.out.println("No Solution ");
				    }                                                                                
				    // restore initial state                                                         
				    chocoModel.getEnvironment().worldPopUntil(0);                                         
				    // remove all constraints relative to user's request                             
				    temp.forEach(chocoModel::unpost);                                                     
				    temp.clear();         
`
Is this the correct way to do it? Also I am using choco-solver-4.0.4

**Comment by cprudhom** [comment]
Personally, I always use the last version of choco, here 4.0.5.

> Is there a way in choco to create these requests array or I have to do it by code?

I don't know how users give you specifications neither how you transformed them into constraints.
Previously, you showed some code with a `HashMap` which seems to be ok to turn requests into constraints. Everything that is not related to choco API can be done as you like.
The code I gave you is a proof of concept:  to show how to backup/restore copy of internal states, how to post/unpost constraints and reset a solver. 

> Is this the correct way to do it?

That seems very close to what I did, so yes I suppose.
Did you execute your code  ? 
Did it work as expected ?
